Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m

```diff
@@ -762,196 +712,275 @@ - (UIFont *)dateHeaderTimeFont
     return [UIFont systemFontOfSize:12.0f];
 }
 
-- (void)loadForTextDisplay
+- (OWSMessageTextView *)createBodyTextViewIfNecessary
 {
-    self.textBubbleImageView.hidden = NO;
-    self.textView.hidden = NO;
-    self.textView.text = self.displayableText.displayText;
-    self.textView.textColor = self.textColor;
+    BOOL shouldIgnoreEvents = NO;
+    if (self.viewItem.interaction.interactionType == OWSInteractionType_OutgoingMessage) {
+        // Ignore taps on links in outgoing messages that haven't been sent yet, as
+        // this interferes with "tap to retry".
+        TSOutgoingMessage *outgoingMessage = (TSOutgoingMessage *)self.viewItem.interaction;
+        shouldIgnoreEvents = outgoingMessage.messageState != TSOutgoingMessageStateSentToService;
+    }
+    [self.class loadForTextDisplay:self.bodyTextViewCached
+                              text:self.displayableBodyText.displayText
+                         textColor:self.textColor
+                              font:self.textMessageFont
+                shouldIgnoreEvents:shouldIgnoreEvents];
+    return self.bodyTextViewCached;
+}
+
++ (void)loadForTextDisplay:(OWSMessageTextView *)textView
+                      text:(NSString *)text
+                 textColor:(UIColor *)textColor
+                      font:(UIFont *)font
+        shouldIgnoreEvents:(BOOL)shouldIgnoreEvents
+{
+    textView.hidden = NO;
+    textView.text = text;
+    textView.textColor = textColor;
 
     // Honor dynamic type in the message bodies.
-    self.textView.font = [self textMessageFont];
-    self.textView.linkTextAttributes = @{
-        NSForegroundColorAttributeName : self.textColor,
+    textView.font = font;
+    textView.linkTextAttributes = @{
+        NSForegroundColorAttributeName : textColor,
         NSUnderlineStyleAttributeName : @(NSUnderlineStyleSingle | NSUnderlinePatternSolid)
     };
-    self.textView.dataDetectorTypes
-        = (UIDataDetectorTypeLink | UIDataDetectorTypeAddress | UIDataDetectorTypeCalendarEvent);
+    textView.dataDetectorTypes = (UIDataDetectorTypeLink | UIDataDetectorTypeAddress | UIDataDetectorTypeCalendarEvent);
+    textView.shouldIgnoreEvents = shouldIgnoreEvents;
+}
 
-    if (self.viewItem.interaction.interactionType == OWSInteractionType_OutgoingMessage) {
-        // Ignore taps on links in outgoing messages that haven't been sent yet, as
-        // this interferes with "tap to retry".
-        TSOutgoingMessage *outgoingMessage = (TSOutgoingMessage *)self.viewItem.interaction;
-        self.textView.shouldIgnoreEvents = outgoingMessage.messageState != TSOutgoingMessageStateSentToService;
+- (BOOL)hasTapForMore
```

Comment:
This class is pretty complicated, so I'm trying to use this kind of utility method to make things more readable and to DRY.